### PR TITLE
Add DeiT (PyTorch)

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds [DeiT (Data-efficient Image Transformers)](https://ai.facebook.com/blog/data-efficient-image-transformers-a-promising-new-technique-for-image-classification/) by Facebook AI.

It's a near-copy of ViT, with only some slight changes:

- DeiT add a distillation token, besides the class token, to the patch sequences to effectively learn from a teacher (CNN).
- Fine-tuning DeiT models can happen in two ways: either as is done in ViT, by only placing a linear layer on top of the final hidden state of the class token, or with a teacher (by placing a linear layer on top of both the final hidden states of the class and distillation tokens). I've defined both models as `DeiTForImageClassification` and `DeiTForImageClassificationWithTeacher`. Currently, the latter supports inference-only. I think that fine-tuning this model with a teacher is something than can be placed in the `examples/research_projects` directory. 
- `DeiTFeatureExtractor` is a bit different compared to `ViTFeatureExtractor`. It uses a different interpolation scheme when resizing images, uses the ImageNet mean/std when normalizing and uses center cropping. 

To do:

- [x] Upload remaining DeiT distilled checkpoints, once `DeiTForImageClassificationWithTeacher` design has been approved.
- [x] Fix integration tests, based on those.
- [x] Update `DeiTFeatureExtractor`, based on functions defined in `image_utils.py`.
- [x] Add `DeiTForImageClassificationWithTeacher` to `modeling_auto.py`. However, we have 2 models with the same key (`DeiTConfig`) in an `OrderedDict`..
- [x] Add model cards.

cc @sgugger @LysandreJik @patil-suraj

### CODE REVIEW

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -579,6 +582,14 @@
             "DebertaV2PreTrainedModel",
         ]
     )
+    _import_structure["models.deit"].extend(
+        [
+            "DEIT_PRETRAINED_MODEL_ARCHIVE_LIST",
+            "DeiTForImageClassification",
+            "DeiTForImageClassificationWithTeacher",
+            "DeiTModel",
+        ]
```
Comment (sgugger) [inline src/transformers/__init__.py:591]: This one is missing `DeiTPreTrainedModel` (or the TPYE_CHECKING part should not have it, whichever makes more sense).

File: src/transformers/models/auto/modeling_auto.py [hunk 2]
```diff
@@ -97,6 +97,9 @@
     DebertaV2ForTokenClassification,
     DebertaV2Model,
 )
+
+# Add modeling imports here
```
Comment (sgugger) [inline src/transformers/models/auto/modeling_auto.py:101]: Should go back on the top.

File: src/transformers/models/deit/modeling_deit.py [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+
+DEIT_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "facebook/deit-base-distilled-patch16-224",
+    # See all DeiT models at https://huggingface.co/models?filter=deit
+]
+
+
+# Inspired by
+# https://github.com/rwightman/pytorch-image-models/blob/b9bd960a032c75ca6b808ddeed76bee5f3ed4972/timm/models/layers/helpers.py
+# From PyTorch internals
```
Comment (sgugger) [inline src/transformers/models/deit/modeling_deit.py:46]: Could we use a `#Copied from modeling.vit` here since it's the same?

File: src/transformers/models/deit/modeling_deit.py [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+        cls_tokens = self.cls_token.expand(batch_size, -1, -1)
+        dist_tokens = self.dist_token.expand(batch_size, -1, -1)
+        embeddings = torch.cat((cls_tokens, dist_tokens, embeddings), dim=1)
+        embeddings = embeddings + self.position_embeddings
+        embeddings = self.dropout(embeddings)
+        return embeddings
+
+
+# Based on timm implementation, which can be found here:
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
```
Comment (sgugger) [inline src/transformers/models/deit/modeling_deit.py:91]: Same here, it looks the same as in ViT.

File: tests/test_modeling_deit.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        # we omit vocab_size since DeiT does not use this
+        self.config_tester.parent.assertTrue(hasattr(config, "hidden_size"))
+        self.config_tester.parent.assertTrue(hasattr(config, "num_attention_heads"))
+        self.config_tester.parent.assertTrue(hasattr(config, "num_hidden_layers"))
+
+        self.config_tester.create_and_test_config_to_json_string()
+        self.config_tester.create_and_test_config_to_json_file()
+        self.config_tester.create_and_test_config_from_and_save_pretrained()
+        self.config_tester.create_and_test_config_with_num_labels()
+        self.config_tester.check_config_can_be_init_without_params()
```
Comment (sgugger) [inline tests/test_modeling_deit.py:171]: This is the same as in ViT and we will have more models like that, so we should add an option in `ConfigTester` to ignore the test of vocab_size, for instance with a flag `has_text_modality`.

**Review by sgugger — APPROVED** [review]
Great PR, thanks a lot! It looks like a lot of the modules are the same as in ViT so they should have a `# Copied from ` flag to make sure they don't diverge. This is going especially useful if we have a bug to fix later on, as it will avoid copying the same thing in all vision modeling files. I've put a comment on the first one, but all the building blocks looked the same apart form the embeddings.

File: tests/test_modeling_deit.py (hunk 5 above)
Comment (NielsRogge) [reply tests/test_modeling_deit.py:171]: Ok, I've added this attribute

**Comment by NielsRogge** [comment]
Thanks for the review, I've addressed your comments.

However, I'm getting the following error when running `make quality`:
```
Exception: Found the following copy inconsistencies:
- src/transformers\models\deit\modeling_deit.py: copy does not match models.vit.modeling_vit.ViTSelfOutput at line 177
- src/transformers\models\deit\modeling_deit.py: copy does not match models.vit.modeling_vit.ViTPreTrainedModel at line 372
```

Even though they are exact copies..

**Comment by sgugger** [comment]
> I see why, the first one has "DeiTLayer" in its docstring instead of "ViTLayer".

Not the Vit one has "VitLayer" which is why the fix-copy fails.

> The other one is not an exact copy as it has "deit" as base_model_prefix instead of "vit". How to resolve this?

You can add `all-casing` at the end of your Copied from statement so it replaces the lowercase version of ViT by DeiT

> Also, I'd like to add a # copied from to DeiTModel, should I place this comment above the @add_start_docstrings annotator?

It should be between the add_start_docstrings and the class.

**Comment by NielsRogge** [comment]
Thanks, `make fix-copies` is ok now. 

@LysandreJik let me know if you agree on the design of `DeiTForImageClassificationWithTeacher` (which raises a `NotImplementedError` if one specifies labels since this model supports inference only), then I can finish this PR today.

File: src/transformers/models/deit/modeling_deit.py [hunk 6]
```diff
@@ -0,0 +1,736 @@
...
+        dist_logits = self.dist_classifier(sequence_output[:, 1, :])
+
+        # during inference, return the average of both classifier predictions
+        logits = (cls_logits + dist_logits) / 2
+
+        loss = None
+        if labels is not None:
+            # TODO add support for fine-tuning with distillation
+            # however this relies on a teacher
+            raise NotImplementedError("Fine-tuning with distillation is not yet implemented.")
```
Comment (NielsRogge) [inline src/transformers/models/deit/modeling_deit.py:725]: DeiTForImageClassificationWithTeacher is an inference-only model.

File: src/transformers/models/deit/modeling_deit.py [hunk 7]
```diff
@@ -0,0 +1,736 @@
...
+        super().__init__(config)
+
+        self.num_labels = config.num_labels
+        self.deit = DeiTModel(config, add_pooling_layer=False)
+
+        # Classifier heads
+        self.cls_classifier = (
+            nn.Linear(config.hidden_size, config.num_labels) if config.num_labels > 0 else nn.Identity()
+        )
+        self.dist_classifier = (
```
Comment (patil-suraj) [inline src/transformers/models/deit/modeling_deit.py:659]: (nit) `dist` is not informative, maybe `distilled_classifier` ?

File: src/transformers/models/deit/modeling_deit.py [hunk 8]
```diff
@@ -0,0 +1,736 @@
...
+            head_mask=head_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        sequence_output = outputs[0]
+
+        cls_logits = self.cls_classifier(sequence_output[:, 0, :])
+        dist_logits = self.dist_classifier(sequence_output[:, 1, :])
```
Comment (patil-suraj) [inline src/transformers/models/deit/modeling_deit.py:716]: same comment as above

**Review by patil-suraj — APPROVED** [review]
LGTM, very clean PR! Thanks a lot for adding this model.

Just left a couple of nits.

File: docs/source/model_doc/deit.rst [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+expensive infrastructure, thereby limiting their adoption. In this work, we produce a competitive convolution-free
+transformer by training on Imagenet only. We train them on a single computer in less than 3 days. Our reference vision
+transformer (86M parameters) achieves top-1 accuracy of 83.1% (single-crop evaluation) on ImageNet with no external
+data. More importantly, we introduce a teacher-student strategy specific to transformers. It relies on a distillation
+token ensuring that the student learns from the teacher through attention. We show the interest of this token-based
+distillation, especially when using a convnet as a teacher. This leads us to report results competitive with convnets
+for both Imagenet (where we obtain up to 85.2% accuracy) and when transferring to other tasks. We share our code and
+models.*
+
+Tips:
```
Comment (LysandreJik) [inline docs/source/model_doc/deit.rst:41]: As always, fantastic work on adding all of those tips @NielsRogge!

File: src/transformers/models/deit/configuration_deit.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    # See all DeiT models at https://huggingface.co/models?filter=deit
+}
+
+
+class DeiTConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a :class:`~transformers.DeiTModel`. It is used to
+    instantiate an DeiT model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the DeiT
+    `facebook/deit-base-distilled-patch16-224 <https://huggingface.co/facebook/deit-base-distilled-patch16-2244>`__
```
Comment (LysandreJik) [inline src/transformers/models/deit/configuration_deit.py:34]:
```suggestion
    `facebook/deit-base-distilled-patch16-224 <https://huggingface.co/facebook/deit-base-distilled-patch16-224>`__
```

File: src/transformers/models/deit/feature_extraction_deit.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
+# coding=utf-8
+# Copyright Google AI and The HuggingFace Inc. team. All rights reserved.
```
Comment (LysandreJik) [inline src/transformers/models/deit/feature_extraction_deit.py:2]:
```suggestion
# Copyright 2021 The HuggingFace Inc. team. All rights reserved.
```

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py [hunk 12]
```diff
@@ -151,23 +151,37 @@ def convert_vit_checkpoint(vit_name, pytorch_dump_folder_path):
...
-        config.hidden_size = 1024
-        config.intermediate_size = 4096
-        config.num_hidden_layers = 24
-        config.num_attention_heads = 16
-    elif vit_name[4:].startswith("huge"):
-        config.hidden_size = 1280
-        config.intermediate_size = 5120
-        config.num_hidden_layers = 32
-        config.num_attention_heads = 16
+    if "deit" in vit_name:
```
Comment (LysandreJik) [inline src/transformers/models/vit/convert_vit_timm_to_pytorch.py:154]: Should DeiT be handled by the ViT conversion script? I would put that in the DeiT conversion script aboee

**Review by LysandreJik — APPROVED** [review]
Great, this LGTM! I left a few comments, but appart from that it looks good to merge for me! Thanks for your work @NielsRogge, impressive once again!

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py (hunk 12 above)
Comment (NielsRogge) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:154]: The story is a bit complicated: deit has checkpoints which are just more efficient ViT models, hence I converted those using the original ViT conversion script (and added "deit" there), and there are also distilled checkpoints, which are converted in `convert_deit_timm_to_pytorch`.py.

Comment (LysandreJik) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:154]: Hmmm okay, as long as that's very well documented that works for me

File: src/transformers/models/deit/modeling_deit.py [hunk 13]
```diff
@@ -0,0 +1,770 @@
+# coding=utf-8
+# Copyright 2021 Facebook AI Research (FAIR), Ross Weightman, The HuggingFace Inc. team. All rights reserved.
```
Comment (LysandreJik) [inline src/transformers/models/deit/modeling_deit.py:2]:
```suggestion
# Copyright 2021 Facebook AI Research (FAIR), Ross Wightman, The HuggingFace Inc. team. All rights reserved.
```
